### PR TITLE
Add a default mapping from EventLoopGroup and ChannelFactory #6839

### LINKED ISSUE #6839: Add a simple way to fallback to Nio if Epoll is not avaiable
This is just a wish, Netty is exciting but I would like to make it more "confortable"

### Expected behavior
Actually there is some boilerplate code to write down if you want to use Epoll and fallback to Nio if Epoll is not available.
I would like to have some Netty-provided facility to boot with Epoll but automatically falls back to Nio 

Examples:
you have to create a EpollEventLoopGroup and fallback to a NioEventLoopGroup in case of ExceptionInInitializerError or NoClassDefFoundError

Then you have to write code like
this:
 ```
Bootstrap b = new Bootstrap();
        b.group(group)
         .channel(nio ? NioSocketChannel.class : EpollSocketChannel.class)
```

or
 ```
Bootstrap b = new Bootstrap();
        b.group(group)
         .channel(group instanceof  NioEventLoopGroup ? NioSocketChannel.class : EpollSocketChannel.class)
```

I see (i.e. I am using networkless DefaultEventLoopGroup too) that there is a tight coupling between the Channel class and the EventLoopGroup class, it would be great that Netty could have an automatic mapping for such configurations.  

### Netty version
4.1.9Final

### PR DESCRIPTION
**Pull request by eolivelli**
Introduce a EventLoopGroup#defaultChannelFactory as an hint to build a ChannelFactory if not configured.
This way the Channel type will be automatically discovered for simple use-cases

Motivation:

see issue #6839 

Modification:

I have added a EventLoopGroup#defaultChannelFactory(boolean serverSide) method to let the EventLoopGroup describe the usual type of Channel it handles.

This way code will be simpler:

before the change:
```
 ServerBootstrap b = new ServerBootstrap();
            b.group(bossGroup, workerGroup)
                .channel(NetworkUtils.isEnableEpoolNative() ? EpollServerSocketChannel.class : NioServerSocketChannel.class)
                .childHandler(channelInitialized)
                .option(ChannelOption.SO_BACKLOG, 128);

```
After the change:
```
 ServerBootstrap b = new ServerBootstrap();
            b.group(bossGroup, workerGroup)
                .childHandler(channelInitialized)
                .option(ChannelOption.SO_BACKLOG, 128);

```

Result:

Fixes #6839 

I have some questions:
- I needed to change the 'validate' method in order to create the channelFactory if not given, this is not very good for me but I did not find another way
- this PR does not include new test cases, I will add them as soon as the general approach to the problem wil be approved (at the moment I have tested all netty based projects and all is running as expected)

### CODE REVIEW

File: transport/src/main/java/io/netty/channel/EventLoopGroup.java [hunk 1]
```diff
@@ -49,4 +49,10 @@
      */
     @Deprecated
     ChannelFuture register(Channel channel, ChannelPromise promise);
+
+    /**
+     * Return the defaults {@link ChannelFactory} for channels to be created in the context of this group
+     * @return a default channel factory or null if not provided
+     */
+    ChannelFactory defaultChannelFactory(boolean serverSide);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/EventLoopGroup.java]: @eolivelli we can not do this for now as this would be a breaking change (adding a method to an interface

Comment (eolivelli) [inline transport/src/main/java/io/netty/channel/EventLoopGroup.java]: @normanmaurer that you for your review

I understand this point.
Netty is compiled for Java6 (I think) and so defaults method are not supported.
Another way to implement this feature would be to create a global registry which maps EventLoopGroup class with the Channel class.
This registry will be filled in static {} blocks in EventLoopGroup subclasses

I can file a new PR in order to demostrate this alternative approach

File: transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueEventLoop.java [hunk 2]
```diff
@@ -15,8 +15,10 @@
  */
 package io.netty.channel.kqueue;
 
+import io.netty.channel.ChannelFactory;
 import io.netty.channel.EventLoop;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ReflectiveChannelFactory;
```
Comment (netkins) [inline transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedEventLoop.java [hunk 3]
```diff
@@ -16,11 +16,13 @@
 package io.netty.channel.embedded;
 
 import io.netty.channel.Channel;
+import io.netty.channel.ChannelFactory;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelPromise;
 import io.netty.channel.DefaultChannelPromise;
 import io.netty.channel.EventLoop;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ReflectiveChannelFactory;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/DefaultEventLoop.java [hunk 4]
```diff
@@ -15,6 +15,8 @@
  */
 package io.netty.channel;
 
+import io.netty.channel.local.LocalChannel;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.local.LocalChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/DefaultEventLoop.java [hunk 5]
```diff
@@ -15,6 +15,8 @@
...
+import io.netty.channel.local.LocalServerChannel;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.local.LocalServerChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java [hunk 6]
```diff
@@ -17,10 +17,14 @@
 
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
+import io.netty.channel.ChannelFactoriesRegistry;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.ChannelFactoriesRegistry'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java [hunk 7]
```diff
@@ -17,10 +17,14 @@
...
 import io.netty.channel.EventLoop;
 import io.netty.channel.EventLoopException;
+import io.netty.channel.ReflectiveChannelFactory;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java [hunk 8]
```diff
@@ -17,10 +17,14 @@
...
 import io.netty.channel.SelectStrategy;
 import io.netty.channel.SingleThreadEventLoop;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.socket.nio.NioServerSocketChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java [hunk 9]
```diff
@@ -17,10 +17,14 @@
...
+import io.netty.channel.socket.nio.NioSocketChannel;
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.socket.nio.NioSocketChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+package io.netty.channel;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * This class provides a default mapping between {@link EventLoopGroup} implementations and {@link ChannelFactory}
+ * implementations
+ */
+public abstract class ChannelFactoriesRegistry {
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java [hunk 11]
```diff
@@ -218,6 +220,14 @@ public B validate() {
         return (B) this;
     }
 
+    private void ensureChannelFactory() {
+        if (channelFactory == null) {
+            channelFactory = ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), isServerSide());
+            System.out.println("[DEBUG] channelFactory for " + group.getClass()
```
Comment (netkins) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: transport/src/main/java/io/netty/channel/DefaultEventLoop.java (hunk 4 above)
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.local.LocalChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/DefaultEventLoop.java (hunk 5 above)
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.local.LocalServerChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java (hunk 6 above)
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.ChannelFactoriesRegistry'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java (hunk 7 above)
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.ReflectiveChannelFactory'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java (hunk 8 above)
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.socket.nio.NioServerSocketChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/nio/NioEventLoop.java (hunk 9 above)
Comment (netkins) [inline transport/src/main/java/io/netty/channel/nio/NioEventLoop.java]: MINOR Remove this unused import 'io.netty.channel.socket.nio.NioSocketChannel'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java (hunk 10 above)
Comment (netkins) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java (hunk 11 above)
Comment (netkins) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

**Comment by mp911de** [comment]
It would make sense to include `KQueue` as well.

**Comment by eolivelli** [comment]
@mp911de I pushed the support for KQueue

@normanmaurer If the general approch is good I would like to add test cases.
Which would be the best place to put them ?
I think that as I am adding default socket channel configuration for Nio, Default, Epoll and KQueue at least I should add a test case in any of the projects

File: transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java [hunk 12]
```diff
@@ -499,4 +509,6 @@ protected EventExecutor executor() {
             return GlobalEventExecutor.INSTANCE;
         }
     }
+
+    protected abstract boolean isServerSide();
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: This method cannot be abstract because this class is public and it would break the API contract.

File: transport/src/main/java/io/netty/channel/ThreadPerChannelEventLoop.java [hunk 13]
```diff
@@ -93,4 +93,5 @@ protected void deregister() {
         parent.activeChildren.remove(this);
         parent.idleChildren.add(this);
     }
+
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/ThreadPerChannelEventLoop.java]: Please revert back unrelated formatting changes

File: transport/src/test/java/io/netty/channel/local/LocalChannelTest.java [hunk 14]
```diff
@@ -21,6 +21,7 @@
 import io.netty.buffer.Unpooled;
 import io.netty.channel.AbstractChannel;
 import io.netty.channel.Channel;
+import io.netty.channel.ChannelFactory;
```
Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/local/LocalChannelTest.java]: Revert please

**Comment by eolivelli** [comment]
@Scottmitch 
I have implemented only the cases for which I am sure about the most convenient default channel type.
My experience is only limited to "tcp" on Default/Nio/Epoll. I have added KQueue as requested.

Should I add some more channel types ?
Could you please give me an hint ?

**Comment by eolivelli** [comment]
@Scottmitch  @normanmaurer 
does the current approch for you ?

@Scottmitch could you please give me some hint ? (see my answer to your change request)

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+    private static ConcurrentMap<Class<? extends EventLoopGroup>, ChannelFactory> clientSideFactories
+        = new ConcurrentHashMap<Class<? extends EventLoopGroup>, ChannelFactory>();
+
+    private static ConcurrentMap<Class<? extends EventLoopGroup>, ChannelFactory> serverSideFactories
+        = new ConcurrentHashMap<Class<? extends EventLoopGroup>, ChannelFactory>();
+
+    private ChannelFactoriesRegistry() { }
+
+    public static void registerFactoryForEventLoopGroup(Class<? extends EventLoopGroup> groupClass,
```
Comment (rkapsi) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: This is possibly not good for forward compatibility. Multiple people have filed issues over `EpollEventLoopGroup`, `NioEventLoopGroup` and so forth being concrete+final classes and basically asked them to be interfaces.

If we were to do that then this approach would no longer work. IMHO it'd be better to use `instanceof` in some shape or form instead of depending on hashCode/equals.

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * This class provides a default mapping between {@link EventLoopGroup} implementations and {@link ChannelFactory}
+ * implementations
+ */
+public abstract class ChannelFactoriesRegistry {
+
+    private static ConcurrentMap<Class<? extends EventLoopGroup>, ChannelFactory> clientSideFactories
```
Comment (rkapsi) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: `final`, upper case, just `Map` on the left hand side, I believe there is a factory method for concurrent maps in the `PlatformDependent` class or so.

**Comment by eolivelli** [comment]
@rkapsi I tried to address your request. 
I have added a test case for the case of using EventLoopGroup subclasses

File: transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java [hunk 17]
```diff
@@ -218,6 +220,12 @@ public B validate() {
         return (B) this;
     }
 
+    private void ensureChannelFactory() {
+        if (channelFactory == null) {
+            channelFactory = ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), isServerSide());
```
Comment (rkapsi) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: Wonder if you can get away with `(this instanceof ServerBootstrap)` instead of the `isServerSide()` method.

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (rkapsi) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: 2017

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+ */
+public abstract class ChannelFactoriesRegistry {
+
+    private static final class ChannelFactoryDefaults {
+        private final Class<? extends EventLoopGroup> eventLoopGroupType;
+        private final ChannelFactory clientSideFactory;
+        private final ChannelFactory serverSideFactory;
+
+        public ChannelFactoryDefaults(Class<? extends EventLoopGroup> eventLoopGroupType,
+            ChannelFactory clientSideFactory, ChannelFactory serverSideFactory) {
```
Comment (rkapsi) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: nit: I believe this needs an extra indent and I believe (unfortunately) is to actually align the args. Please double check and/or somebody can confirm.

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    }
+
+    private static final List<ChannelFactoryDefaults> DEFAULT_CHANNEL_FACTORIES
+        = new ArrayList<ChannelFactoryDefaults>();
+
+    private ChannelFactoriesRegistry() { }
+
+    public static void registerFactoryForEventLoopGroup(Class<? extends EventLoopGroup> groupClass,
+        ChannelFactory factoryForClients, ChannelFactory factoryForServers) {
```
Comment (rkapsi) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: Same with these args

File: transport/src/test/java/io/netty/channel/ChannelFactoriesRegistryTest.java [hunk 21]
```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2015 The Netty Project
```
Comment (rkapsi) [inline transport/src/test/java/io/netty/channel/ChannelFactoriesRegistryTest.java]: 2017

**Comment by rkapsi** [comment]
LGTM except for one aspect and something I've struggled myself with the API. The type of the `EventLoop[Group]` doesn't really indicate which "transport layer" the user intends to use. This impl. assumes the user wants to use TCP but it could be UDP, Unix Domain Socket or something else and I don't know if it's a good idea to make such default choice.

We did for a while the opposite and had an `EventLoopGroupManager` class with a `get(Class<? extends Channel> type)` method that provided appropriate `EventLoopGroup` instances for a given channel type and a custom wrapper around Bootstrap what would fill in the group if it was missing. It was OK but had its problems too when it came down to debugging things.

File: transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java (hunk 17 above)
Comment (eolivelli) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: @rkapsi IMHO using "instanceof" in a parent class on 'this' is not a good object oriented design choice.
But I will change it if you confirm it is better for the rules of the project

**Comment by eolivelli** [comment]
@rkapsi thank you
I have fixed the indentation/copyright issues (I followed the example of PooledByteBufAllocator for the indentation).
I will commit the 'instanceof' change if needed, see my comments.

For the question about the support of connection less protocols or Unix Domain Socket....
I think that this feature makes the life simpler to new Netty users for the basic TCP usage, more advanced cases are still possible using the usual API.

File: transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java (hunk 17 above)
Comment (fenik17) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: @eolivelli +1. A class reference to its subclasses is anti-pattern..

Comment (rkapsi) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: How about this...

```java
class AbstractBootstrap {
  private void ensureChannelFactory() {
    if (chanelFactory == null) {
       chanelFactory = getDefaultChannelFactory();
    }
  }

  protected ChannelFactory getDefaultChannelFactory() {
    return null;
  }
}

class Bootstrap extends AbstractBootstrap {
  @Overrides
  protected ChannelFactory getDefaultChannelFactory() {
    return ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), false);
  }
}

class ServerBootstrap extends AbstractBootstrap {
  @Overrides
  protected ChannelFactory getDefaultChannelFactory() {
    return ChannelFactoriesRegistry.getFactoryForEventLoopGroup(group.getClass(), true);
  }
}
```

Comment (eolivelli) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: @rkapsi for me is OK I will update the patch soon, thanks

Comment (fenik17) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: @rkapsi this is better. One nit:
```protected abstract ChannelFactory getDefaultChannelFactory();```

Comment (eolivelli) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: @fenik17 with cannot make it abstract because we would break existing subclasses

Comment (rkapsi) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: I believe `abstract` would be an API breaking change.

Comment (fenik17) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: yep. I updated comment above. 
If `getDefaultChannelFactory()` returns `null`, this will lead to an runtime NPE in the users subclusses.

Comment (rkapsi) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: Nah, should behave the same as right now saying something along the lines of ChannelFactory being missing.

You wouldn't want to return a client channel factory by default if the user's custom bootstrap is for server channels.

Comment (eolivelli) [inline transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java]: @rkapsi @fenik17  getDefaultChannelFactory() by default will return null as it will return for type of EventLoopGroups which are not pre-registered.
I will update the patch soon

**Comment by eolivelli** [comment]
@rkapsi  @fenik17 pushed the refactor

**Comment by eolivelli** [comment]
@rkapsi @fenik17 @Scottmitch @normanmaurer
any other comments ?

**Comment by eolivelli** [comment]
Ping

File: transport/src/test/java/io/netty/channel/ChannelFactoriesRegistryTest.java [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+            group1Future.await();
+            group2Future.await();
+        }
+    }
+
+    static class TestHandler extends ChannelInboundHandlerAdapter {
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            logger.info(String.format("Received message: %s", msg));
```
Comment (fenik17) [inline transport/src/test/java/io/netty/channel/ChannelFactoriesRegistryTest.java]: ```logger.info("Received message: {}", msg);```

Comment (eolivelli) [reply transport/src/test/java/io/netty/channel/ChannelFactoriesRegistryTest.java]: @fenik17 done

**Comment by eolivelli** [comment]
@rkapsi @fenik17 @Scottmitch @normanmaurer
any other comments ?

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java [hunk 23]
```diff
@@ -315,4 +316,9 @@ final SocketAddress remoteAddress() {
     final AddressResolverGroup<?> resolver() {
         return resolver;
     }
+
+    @Override
+    protected io.netty.channel.ChannelFactory getDefaultChannelFactory() {
```
Comment (johnou) [inline transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: do we need the FQCN?

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+            this.clientSideFactory = clientSideFactory;
+            this.serverSideFactory = serverSideFactory;
+        }
+
+        private boolean matchesClass(Class<? extends EventLoopGroup> clazz) {
+            return eventLoopGroupType.isAssignableFrom(clazz);
+        }
+    }
+
+    private static final List<ChannelFactoryDefaults> DEFAULT_CHANNEL_FACTORIES
```
Comment (johnou) [inline transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: I think a COW would make more sense here so we can get rid of the sync.

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java (hunk 23 above)
Comment (eolivelli) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: @johnou yes, otherwise we are hitting a deprecated method

File: transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java (hunk 24 above)
Comment (eolivelli) [reply transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java]: I think we can't. Otherwise we can't implement correctly registerFactoryForEventLoopGroup.

**Comment by eolivelli** [comment]
@johnou I have answered to all of your comments.
Rebased to current 4.1 branch

**Comment by normanmaurer** [comment]
I think we should revisit the way how we do this kind of stuff in netty 5, but for netty 4 this is too late. Let me close this.

**Comment by eolivelli** [comment]
Sure
